### PR TITLE
High Level API

### LINKED ISSUE #18: Higher-level API for user
Using the API is still very verbose. As a user I mostly care about a couple of top-level resources from the APIs and their `spec` field. I think it would be great if this repo would provide a higher-level API built on the generated one that makes it easier to use. For example

```haskell
let Deployment = ./k8s/api/apps/v1/Deployment
in let metadata = Deployment.defaultMetadata { name = "hello-world" }
in let spec = Deployment.defaultSpec
  { selector = Deployment.defaultSelector { ... }
  , template = Deployment.defaultTemplate
    { metadata = Deployment.defaultMetadata
    , spec = Deployment.defaultPodSpec { ... } : Deployment.PodSpec
    }
  } 
in let deployment = Deployment.default { metadata = metadata, spec = spec }: Deployment.Type
```

In this example we actually make `spec` required because having a deployment without it is meaning less. This also reduces some boilerplate with the `Some` constructor.

You can also get all the default constructors and types from the `Deployment` record.

### PR DESCRIPTION
**Pull request by f-f**
This PR adds a high-level API. Fixes #18.

We add hand-written "package-like" objects for the following entities:
- [x] Deployment
- [x] Service

### CODE REVIEW

**Comment by arianvp** [comment]
My remarks are mostly social, instead of about code

Any idea how we're gonna keep the 'high level api' up to date with the latest k8s api definitions?  K8s has versioning, but we need to have incentive to keep up with the api versions as things progress from `alpha -> beta -> v1`.  I'm wondering how we're gonna solve that.

**Comment by f-f** [comment]
@arianvp very good point, I was considering all the implications of having another API on top of Kube lately.

To answer your point: the part of "specifying the API version for objects" (without breaking compatibility I guess) can be solved by just appending a postfix to the object names (e.g. we now have `Deployment`, if that ever comes we'll have `Deployment.v2.alpha`, and so on). The part of "keeping up with updates" is a little bit trickier and basically relies on human curation.

And the more general point about this is: whatever API we end up making by hand it's bound to be _incomplete_. Being fine with this means that we'll have to accept not covering all usecases in one shot, and advising "falling back on the lower-level API" if needed.

Now, the current iteration in this PR so far is only "wrapping" the current API to cut down on lines of imports, and you still have to compose the Kube objects in a sometimes clumsy way.
I've been working on an abstraction on this (which I think it's closer to what @geigerzaehler meant when opening #18), so that you'll have to "configure" your e.g. Deployment by just making a record with all the config you need:

```haskell
{ name = "foo"
, replicas = 2
, containers =
  [ { container = "foo:bar"
    , ...
    }
  , ...
  ]
, ...
}
```

then merging it with a "default record" (that we'll provide), and passing it to a `mkDeployment` function that will actually generate all the boring plumbing.

To sum up my intent: this is going to be incomplete, so I expect to iterate quickly on it as people add their use-cases to it; my priority here is mostly UX.

**Comment by f-f** [comment]
@arianvp @geigerzaehler 
So I just pushed a "higher-level" API over `Deployment`, that hides away a whole bunch of complexity. 

Background: I had a realization from @ryota-ka's [excellent blog post](https://ryota-ka.hatenablog.com/entry/2018/08/27/110000) (thanks for that!) in which he implements a richer version of the boilerplate we have in our examples, and a slightly less complex version of what we have in our `infra` codebase at work to generate the deployments.
Now, I think this boilerplate is general enough so that it can be taken care of by this repo, given that we provide enough configurability. And in the worst case one can always open an issue here and in the meanwhile fall back on the low-level API.

So the new `mkDeployment` function is supposed to take care of the boilerplate, and to generate a Deployment now you just need to write something like:

```haskell
let config =
  ../api/Deployment/default
  //
  { name = "nginx"
  , replicas = 2
  , containers =
    [ ../api/Deployment/defaultContainer
      //
      { name = "nginx"
      , imageName = "nginx"
      , imageVersion = "1.5.3"
      , port = [ 80 ] : Optional Natural
      }
    ]
  }

in ../api/Deployment/mkDeployment config
```

The YAML that it generates is more verbose than the standard basic Kubernetes Deployment example; however this is kind of "production ready" (the aim is to provide a bunch of "sane defaults"):

```yaml
apiVersion: apps/v1
kind: Deployment
spec:
  revisionHistoryLimit: 20
  selector:
    matchLabels:
      app: nginx
  strategy:
    rollingUpdate:
      maxSurge: 5
      maxUnavailable: 0
    type: RollingUpdate
  template:
    spec:
      containers:
      - image: nginx:1.5.3
        imagePullPolicy: Always
        env: []
        resources:
          limits:
            cpu: 500m
          requests:
            cpu: 10m
        name: nginx
        ports:
        - containerPort: 80
      volumes: []
    metadata:
      name: nginx
      labels:
        app: nginx
  replicas: 2
metadata:
  name: nginx
```

If this looks good to you all I'll proceed with finishing it up:
- [x] polish the `mkDeployment`, add some missing things (e.g. `volumeMount`), comments, refactor, etc
- [x] add a test for the `nginxDeployment.dhall`
- [x] repeat the process for `Service` and write a `mkService`, tests, etc
- [x] update the README with describing the "high level" and "low level" APIs

**Comment by f-f** [comment]
@geigerzaehler @arianvp 
The code changes are done (tests are failing because of README build).
I implemented only `Deployment` and `Service` for now (so we can evaluate how good is this), other stuff can go in another PR.

So I'll now fix the README, and merge this in some days if there are no further comments
